File: examples/OCaml/anpr/lib/tsanpr.ml
(* The MIT License (MIT)
   Copyright © 2022-2025 TS-Solution Corp.

   Permission is hereby granted, free of charge, to any person obtaining a copy
   of this software and associated documentation files (the "Software"), to deal
   in the Software without restriction, including without limitation the rights
   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
   copies of the Software, and to permit persons to whom the Software is
   furnished to do so, subject to all conditions.

   The above copyright notice and this permission notice shall be included in all
   copies or substantial portions of the Software.

   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
   SOFTWARE. *)

open Ctypes
open Foreign

type t = {
  lib : Dl.library;
  anpr_initialize : string -> string;
  anpr_read_file : string -> string -> string -> string;
  anpr_read_pixels : unit ptr -> int -> int -> int -> string -> string -> string -> string;
}

let create library_path =
  let lib = Dl.dlopen ~filename:library_path ~flags:[Dl.RTLD_NOW] in
  let anpr_initialize = foreign ~from:lib "anpr_initialize" (string @-> returning string) in
  let anpr_read_file = foreign ~from:lib "anpr_read_file" (string @-> string @-> string @-> returning string) in
  let anpr_read_pixels = foreign ~from:lib "anpr_read_pixels"
    (ptr void @-> int @-> int @-> int @-> string @-> string @-> string @-> returning string) in
  { lib; anpr_initialize; anpr_read_file; anpr_read_pixels }

let anpr_initialize t mode =
  t.anpr_initialize mode

let anpr_read_file t img_file_name output_format options =
  t.anpr_read_file img_file_name output_format options

let anpr_read_pixels t pixels width height stride pixel_format output_format options =
  let ptr = to_voidp (Ctypes.bigarray_start Ctypes.array1 pixels) in
  t.anpr_read_pixels ptr width height stride pixel_format output_format options

let anpr_read_pixels_encoded t encoded_data len output_format options =
  let ptr = to_voidp (Ctypes.bigarray_start Ctypes.array1 encoded_data) in
  t.anpr_read_pixels ptr len 0 0 "encoded" output_format options

File: examples/Nim/anpr/src/tsanpr.nim
# The MIT License (MIT)
# Copyright © 2022-2025 TS-Solution Corp.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to all conditions.
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import dynlib, os

type
  TSANPR* = ref object
    lib: LibHandle

proc cleanupTSANPR(tsanpr: TSANPR) =
  ## Destructor - unload library when object is destroyed
  if tsanpr.lib != nil:
    unloadLib(tsanpr.lib)

proc newTSANPR*(libraryPath: string): TSANPR =
  ## Initialize TSANPR with the given library path.
  if not fileExists(libraryPath):
    raise newException(IOError, "Library file not found: " & libraryPath)

  let lib = loadLib(libraryPath)
  if lib == nil:
    raise newException(OSError, "Failed to load native library: " & libraryPath)

  new(result, cleanupTSANPR)
  result.lib = lib

proc anprInitialize*(tsanpr: TSANPR, mode: string): string =
  ## Initialize the ANPR engine with the specified mode.
  try:
    let anpr_init = cast[proc(mode: cstring): cstring {.cdecl.}](tsanpr.lib.symAddr("anpr_initialize"))
    if anpr_init == nil:
      return "Function anpr_initialize not found"
    
    let result_ptr = anpr_init(mode.cstring)
    if result_ptr == nil:
      return ""
    else:
      return $result_ptr
  except:
    return "Function call failed: " & getCurrentExceptionMsg()

proc anprReadFile*(tsanpr: TSANPR, imgFileName: string, outputFormat: string, options: string): string =
  ## Read and process an image file.
  try:
    let anpr_read = cast[proc(imgFileName: cstring, outputFormat: cstring, options: cstring): cstring {.cdecl.}](
      tsanpr.lib.symAddr("anpr_read_file"))
    if anpr_read == nil:
      return "Function anpr_read_file not found"
    
    let result_ptr = anpr_read(imgFileName.cstring, outputFormat.cstring, options.cstring)
    if result_ptr == nil:
      return ""
    else:
      return $result_ptr
  except:
    return "Function call failed: " & getCurrentExceptionMsg()

proc anprReadPixels*(tsanpr: TSANPR, pixels: ptr UncheckedArray[uint8], width: uint64, 
                    height: uint64, stride: int64, pixelFormat: string, 
                    outputFormat: string, options: string): string =
  ## Process pixel data directly.
  try:
    let anpr_pixels = cast[proc(pixels: ptr UncheckedArray[uint8], width: uint64, height: uint64, 
                               stride: int64, pixelFormat: cstring, outputFormat: cstring, 
                               options: cstring): cstring {.cdecl.}](
      tsanpr.lib.symAddr("anpr_read_pixels"))
    if anpr_pixels == nil:
      return "Function anpr_read_pixels not found"
    
    let result_ptr = anpr_pixels(pixels, width, height, stride, pixelFormat.cstring, 
                                outputFormat.cstring, options.cstring)
    if result_ptr == nil:
      return ""
    else:
      return $result_ptr
  except:
    return "Function call failed: " & getCurrentExceptionMsg()